### ISSUE TITLE
[C++] Unable to insert uprobe at arbitrary file offset?

### ISSUE THREAD

**Issue by ang1337** [iovisor/bcc#3219]
My purpose is to dynamically instrument a stripped binary with an ability to insert uprobes at arbitrary file offsets (that are aligned to instructions of course, not completely arbitrary).

I've tried to use ```attach_uprobe``` method from C++ API, but it seems to work only when a binary does have symbols so I can provide a symbol name, whereas Python API seems to support arbitrary offset probing and thus works well with stripped binaries. Do I try to use a wrong method for that purpose or C++ API simply doesn't support it? Thank you in advance.

**Comment by yonghong-song** [comment]
I think python and C++ you can do with symbol_name+symbol_offset or address. If C++ cannot do "address" only, maybe we have a bug there. Is this what you mean?

**Comment by ang1337** [comment]
> I think python and C++ you can do with symbol_name+symbol_offset or address. If C++ cannot do "address" only, maybe we have a bug there. Is this what you mean?

Yes, C++ API seems not to have an option to provide only address or only file offset option, so dynamic instrumentation of a stripped binary via C++ API seems impossible right now.

BTW, is uprobe amount in one process limited? If yes, can it be manually configured to have a higher limit?

**Comment by yonghong-song** [comment]
Maybe there is a bug? For `attach_uprobe` API, you should be able to specify a symbol address and attach it. The parameters in python API and C++ API are mostly compatible. Could you help debug this?

> BTW, is uprobe amount in one process limited? If yes, can it be manually configured to have a higher limit?
No uprobe attachment is not tied to a particular pid. Once uprobe is attached, the bpf program should be triggered
for different tasks.

**Comment by ang1337** [comment]
> Maybe there is a bug? For `attach_uprobe` API, you should be able to specify a symbol address and attach it. The parameters in python API and C++ API are mostly compatible. Could you help debug this?
> 
Once again, the problem is with stripped binaries. There are no symbols in stripped binaries, so providing a symbol address is useless - there are no symbols at all. My purpose is to be able to set uprobes at arbitrary raw address having zero information about symbols. 
Here is ```attach_uprobe``` function signature in C++ API:
```  StatusTuple attach_uprobe(const std::string& binary_path,
                                                  const std::string& symbol,
                                                  const std::string& probe_func,
                                                  uint64_t symbol_addr = 0,
                                                  bpf_probe_attach_type attach_type = BPF_PROBE_ENTRY,
                                                  pid_t pid = -1,
                                                  uint64_t symbol_offset = 0,
                                                  uint32_t ref_ctr_offset = 0);
```
Pay attention to the second argument ```symbol```. It cannot be ignored, when I try to pass an empty string (there are no symbols in stripped binary, so I cannot provide any symbol information), a runtime error message is popped out: ```Cannot find a symbol```.

> No uprobe attachment is not tied to a particular pid. Once uprobe is attached, the bpf program should be triggered
for different tasks.
>
That's not exactly what I've asked. I mean, if I instrument ONE particular process (according to API, I have an option to specify a particular PID), do I have a limited amount of uprobes that can be inserted in one particular process?

**Comment by yonghong-song** [comment]
> Pay attention to the second argument symbol. It cannot be ignored, when I try to pass an empty string (there are no symbols in stripped binary, so I cannot provide any symbol information), a runtime error message is popped out: Cannot find a symbol.

We probably have a bug here. If you just specify symbol_addr (similar to python), bcc also tries to attach to the file offset corresponding to that symbol address. Do you want to take a deep look?

> I mean, if I instrument ONE particular process (according to API, I have an option to specify a particular PID), do I have a limited amount of uprobes that can be inserted in one particular process?

There is no limits unless you run out of memory, file descriptors, etc.
